pkg/instance/instance.go

```diff
@@ -0,0 +1,588 @@
+	logrus.WithFields(logrus.Fields{
+		"name":               req.Spec.Name,
+		"type":               req.Spec.Type,
+		"backendStoreDriver": req.Spec.BackendStoreDriver,
+	}).Info("Replacing instance")
+
+	switch req.Spec.BackendStoreDriver {
+	case rpc.BackendStoreDriver_longhorn:
+		return s.processInstanceReplace(req)
+	case rpc.BackendStoreDriver_spdk:
+		return s.spdkInstanceReplace(req)
+	default:
+		return nil, grpcstatus.Errorf(grpccodes.InvalidArgument, "unknown backend store driver %v", req.Spec.BackendStoreDriver)
+	}
+}
+
+func (s *Server) processInstanceReplace(req *rpc.InstanceReplaceRequest) (*rpc.InstanceResponse, error) {
+	if req.Spec.ProcessSpecific == nil {
+		return nil, fmt.Errorf("process is required for longhorn backend store driver")
+	}
+
+	pmClient, err := client.NewProcessManagerClient("tcp://"+s.processManagerServiceAddress, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer pmClient.Close()
+
+	process, err := pmClient.ProcessReplace(req.Spec.Name,
+		req.Spec.ProcessSpecific.Binary, int(req.Spec.PortCount), req.Spec.ProcessSpecific.Args, req.Spec.PortArgs, req.TerminateSignal)
+	if err != nil {
+		return nil, err
+	}
+
+	return processResponseToInstanceResponse(process), nil
+}
+
+func (s *Server) spdkInstanceReplace(req *rpc.InstanceReplaceRequest) (*rpc.InstanceResponse, error) {
+	return nil, fmt.Errorf("spdk instance replace is not supported")
+}
+
+func (s *Server) InstanceLog(req *rpc.InstanceLogRequest, srv rpc.InstanceService_InstanceLogServer) error {
+	logrus.WithFields(logrus.Fields{
+		"name":               req.Name,
+		"type":               req.Type,
+		"backendStoreDriver": req.BackendStoreDriver,
+	}).Info("Getting instance log")
+
+	switch req.BackendStoreDriver {
+	case rpc.BackendStoreDriver_longhorn:
+		return s.processInstanceLog(req, srv)
+	case rpc.BackendStoreDriver_spdk:
+		return s.spdkInstanceLog(req, srv)
+	default:
+		return grpcstatus.Errorf(grpccodes.InvalidArgument, "unknown backend store driver %v", req.BackendStoreDriver)
+	}
+}
+
+func (s *Server) processInstanceLog(req *rpc.InstanceLogRequest, srv rpc.InstanceService_InstanceLogServer) error {
+	pmClient, err := client.NewProcessManagerClient("tcp://"+s.processManagerServiceAddress, nil)
+	if err != nil {
+		return err
+	}
+	defer pmClient.Close()
+
+	stream, err := pmClient.ProcessLog(context.Background(), req.Name)
+	if err != nil {
+		return err
+	}
+	for {
+		line, err := stream.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			logrus.WithError(err).Error("Failed to receive log")
+			return err
+		}
+
+		if err := srv.Send(&rpc.LogResponse{Line: line}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Server) spdkInstanceLog(req *rpc.InstanceLogRequest, srv rpc.InstanceService_InstanceLogServer) error {
+	return fmt.Errorf("spdk instance log is not supported")
+}
+
+func (s *Server) InstanceWatch(req *empty.Empty, srv rpc.InstanceService_InstanceWatchServer) error {
```

Comment:
Does this instance server needs to implement the broadcasting mechanism?

Reply:
Not necessary. Any concern?

Reply:
I see. This instance watch implementation blindly starts new connections with the below process & spdk servers then forwards the messages to the caller.
```
monitor thread A in longhorn manager == Instance Server == Process/SPDK Server
monitor thread B in longhorn manager == Instance Server == Process/SPDK Server
```

The ideal implementation is like this, which always uses a single connection between Instance Server and Process/SPDK server to handle all watch streaming:
```
monitor thread A in longhorn manager  
                                                                 \
                                                                   == Instance Server == Process/SPDK Server
                                                                 /
monitor thread B in longhorn manager 
```
But it's a little bit complicated while the gain is little. We can ignore it for now.